banzai/calibrations.py

```diff
@@ -108,37 +108,48 @@ def master_selection_criteria(self):
     def calibration_type(self):
         pass
 
+    def construct_file_info_for_master_calibration(self, master_calibration_image):
+        file_info = {'frameid': master_calibration_image.frameid,
+                     'path': os.path.join(master_calibration_image.filepath, master_calibration_image.filename)}
+        return file_info
+
     def on_missing_master_calibration(self, image):
         msg = 'Master Calibration file does not exist for {stage}, flagging image as bad'
         logger.error(msg.format(stage=self.stage_name), image=image)
         image.is_bad = True
 
     def do_stage(self, image):
-        master_calibration_filename = self.get_calibration_filename(image)
+        master_calibration = self.get_calibration_image(image)
 
-        if master_calibration_filename is None:
+        if master_calibration is None:
             self.on_missing_master_calibration(image)
             return image
 
-        master_calibration_image = FRAME_CLASS(self.runtime_context, filename=master_calibration_filename)
+        file_info = self.construct_file_info_for_master_calibration(master_calibration)
+
+        master_calibration_image = FRAME_CLASS(self.runtime_context, file_info=file_info)
+        # cache the master calibration image for future use
+        if not os.path.isfile(file_info.get('path')):
```

Comment:
I would just put this logic in the constructor rather than here.

Reply:
We have decided to keep the caching logic here, as we do not wish to cache every frame that is downloaded.

The reason for placing the caching logic in the `ApplyCalibration` class is that every time an `ApplyCalibration` stage is run (any stage that utilizes a master calibration image to perform reduction), the master cal is re-opened. When working on `/archive` this wasn't a big deal, but now with the likelihood of downloading the file from s3, we need to be smarter about this as we don't wish to re-download a master calibration image 3-4 times or more per reduction.